packages/beacon-state-transition/src/fast/stateTransition.ts

```diff
@@ -0,0 +1,61 @@
+import {allForks, phase0} from "@chainsafe/lodestar-types";
+import {processBlock, processSlots} from "../phase0/fast";
+import {verifyProposerSignature} from "./signatureSets";
+import {CachedBeaconState} from "./util";
+
+// Multifork capable state transition
+
+/**
+ * Implementation of protolambda's eth2fastspec (https://github.com/protolambda/eth2fastspec)
+ */
+export function fastStateTransition(
+  state: CachedBeaconState<allForks.BeaconState>,
+  signedBlock: allForks.SignedBeaconBlock,
+  options?: {verifyStateRoot?: boolean; verifyProposer?: boolean; verifySignatures?: boolean}
+): CachedBeaconState<allForks.BeaconState> {
+  const {verifyStateRoot = true, verifyProposer = true, verifySignatures = true} = options || {};
+  const {config} = state;
+
+  const preSlot = state.slot;
+  const preFork = config.getForkName(preSlot);
+  const block = signedBlock.message;
+  const blockSlot = block.slot;
+  const blockFork = config.getForkName(blockSlot);
+  const postState = state.clone();
+
+  // process slots (including those with no blocks) since block
+  switch (preFork) {
+    case "phase0":
+      processSlots(postState as CachedBeaconState<phase0.BeaconState>, block.slot);
+      break;
+    default:
+      throw new Error(`Slot processing not implemented for fork ${preFork}`);
+  }
+
+  // perform state upgrades
+  // TODO
+
+  // verify signature
+  if (verifyProposer) {
+    if (!verifyProposerSignature(postState, signedBlock)) {
+      throw new Error("Invalid block signature");
```

Comment:
maybe add block slot/root here

Reply:
Should we use detailed typed errors in the state transition function too?

Reply:
These errors (this and below) are untouched from before. I'll make an issue to revisit state transition errors for a future PR!